### PR TITLE
Add promocode microservice

### PR DESCRIPTION
**Pull request by irnrogalik**

### CODE REVIEW

File: data-source.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { join } from 'path';
+import { DataSource } from 'typeorm';
+import { SnakeNamingStrategy } from 'typeorm-naming-strategies';
+
+export const appDataSource = new DataSource({
+  type: 'postgres',
+  host: 'localhost',
+  port: 5454,
+  username: 'postgres',
+  password: 'postgres',
```
Comment (dianager) [inline data-source.ts]: should be moved to configs

File: src/migrations/1651742798938-promocode_table.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`
+    CREATE TABLE IF NOT EXISTS promocode
+    (
+        "id"        uuid              NOT NULL DEFAULT gen_random_uuid(),
+        "name"      character varying NOT NULL,
+        "percent"   INT               NOT NULL,
+        "isUsed"    BOOLEAN           DEFAULT FALSE,
+        "isOneTime" BOOLEAN           DEFAULT FALSE,
+        "startDate" character varying DEFAULT '',
```
Comment (dianager) [inline src/migrations/1651742798938-promocode_table.ts]: when working with date we need to use timestamp type, if no date exists we need to use null not empty string

File: src/migrations/1651742798938-promocode_table.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+export class promocodeTable1651742798938 implements MigrationInterface {
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`
+    CREATE TABLE IF NOT EXISTS promocode
+    (
+        "id"        uuid              NOT NULL DEFAULT gen_random_uuid(),
+        "name"      character varying NOT NULL,
+        "percent"   INT               NOT NULL,
+        "isUsed"    BOOLEAN           DEFAULT FALSE,
```
Comment (dianager) [inline src/migrations/1651742798938-promocode_table.ts]: do we need this field? We have promocode which is active from startDat to endDate.  When we are going to set this field?

File: src/promo/promo.repository.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+          promocode.startDate,
+          promocode.endDate,
+        ],
+    );
+    return newPromocode[0];
+  }
+
+  async removePromoCode(promocodeName: IPromoCodeName): Promise<Boolean> {
+    const response = await appDataSource.query(
+        'DELETE FROM promocode WHERE name = $1', [promocodeName.name],
```
Comment (dianager) [inline src/promo/promo.repository.ts]: can we add softDelete? Also the reason why promocode was deleted

File: src/migrations/1651742798938-promocode_table.ts (hunk 3 above)
Comment (dianager) [reply src/migrations/1651742798938-promocode_table.ts]: we can change for usedDate or similar filed to know when the promocode was used.

File: src/migrations/1651742798938-promocode_table.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import type { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class promocodeTable1651742798938 implements MigrationInterface {
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`
+    CREATE TABLE IF NOT EXISTS promocode
+    (
+        "id"        uuid              NOT NULL DEFAULT gen_random_uuid(),
+        "name"      character varying NOT NULL,
```
Comment (dianager) [inline src/migrations/1651742798938-promocode_table.ts]: add index for this field

File: src/promo/promo.proto [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    
+message PromoCodeName {
+	string name = 1;
+}
+
+message PromoCodeBoolResponse {
+	bool response = 1;
+}
+
+message GetPromoCodeList {}
```
Comment (dianager) [inline src/promo/promo.proto]: add size, limit options.

File: src/promo/promo.repository.ts (hunk 4 above)
Comment (dianager) [reply src/promo/promo.repository.ts]: deletedAt  deletedBy - add instead of hard delete

File: src/promo/promo.repository.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  async markPromoCodeAsUsed(promocodeName: IPromoCodeName): Promise<Boolean> {
+    const response = await appDataSource.query(
+        'UPDATE promocode SET "isUsed" = true WHERE name = $1 RETURNING *', [promocodeName.name],
+    );
+    return response[0];
+  }
+
+  async getPromoCodeByName(promocodeName: IPromoCodeName): Promise<IPromoCode> {
+    const promocode = await appDataSource.query(
+        'SELECT * FROM promocode where name = $1', [promocodeName.name],
```
Comment (dianager) [inline src/promo/promo.repository.ts]: nit: WHERE not where

File: data-source.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import 'dotenv/config';
+import { join } from 'path';
+import { DataSource } from 'typeorm';
+import { SnakeNamingStrategy } from 'typeorm-naming-strategies';
+
+export const appDataSource = new DataSource({
+  type: 'postgres',
+  host: process.env.DB_HOST,
+  port: +process.env.DB_PORT,
+  username: process.env.DB_USERNAME,
```
Comment (dianager) [inline data-source.ts]: i see you use ConfigModule. We can get all these variables from ConfigService

File: src/promo/entity/promo.entity.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+@Entity({ name: 'promocode' })
+export class PromocodeEntity implements IPromoCode {
+  @PrimaryGeneratedColumn()
+    id: string;
+
+  @Column()
+    name: string;
+
+  @Column()
+    createdAt: string;
```
Comment (dianager) [inline src/promo/entity/promo.entity.ts:13]: typeorm provides autogenerated filed. Here we can use this pattern. We need to add appropriate decorator

File: src/promo/entity/promo.entity.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    usedDate: string;
+
+  @Column()
+    startDate: string;
+
+  @Column()
+    endDate: string;
+
+  @Column()
+    deletedAt: string;
```
Comment (dianager) [inline src/promo/entity/promo.entity.ts]: string | null

File: src/promo/entity/promo.entity.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+  @Column()
+    deletedReason: string;
```
Comment (dianager) [inline src/promo/entity/promo.entity.ts]: string | null

File: src/promo/entity/promo.entity.ts [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    percent: number;
+
+  @Column()
+    currentState: PromocodeState;
+
+  @Column()
+    isOneTime: boolean;
+
+  @Column()
+    usedDate: string;
```
Comment (dianager) [inline src/promo/entity/promo.entity.ts]: as far as i remember we set this field if the code is one time used. that's why string | null

File: src/promo/promo.repository.ts [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+import { appDataSource } from '../../data-source';
+import { PromocodeEntity } from './entity/promo.entity';
+import { IAddPromoCode, IPageOptions, IPromoCode, IPromoCodeName, IRemovePromoCode } from './promo.interface';
+import { PromocodeState } from './promocode.enum';
+
+export class PromocodeRepository {
+  async getListOfPromocodes(pageOptions: IPageOptions): Promise<PromocodeEntity[]> {
+    let query = 'SELECT * FROM promocode';
+
+    if (!pageOptions.page) {
```
Comment (dianager) [inline src/promo/promo.repository.ts]: in promo.proto this field is required. the same about limit. do we need this check?

File: src/promo/entity/promo.entity.ts [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+    name: string;
+
+  @Column()
+    createdAt: string;
+
+  @Column()
+    percent: number;
+
+  @Column()
+    currentState: PromocodeState;
```
Comment (dianager) [inline src/promo/entity/promo.entity.ts]: don't think that we need this field. promocode is active if it's not deleted. we can add check that deletedAt field is null to get al active promocodes.

File: src/promo/promo.repository.ts [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+        ],
+    );
+    return newPromocode[0];
+  }
+
+  async removePromoCode(promocode: IRemovePromoCode): Promise<Boolean> {
+    const response = await appDataSource.query(
+        `UPDATE promocode
+        SET
+          "currentState" = '${ PromocodeState.REMOVED }',
```
Comment (dianager) [inline src/promo/promo.repository.ts]: don't think we need this status. or do you plan to add ned status for promocodes in future?

File: src/promo/promo.repository.ts [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+import { appDataSource } from '../../data-source';
+import { PromocodeEntity } from './entity/promo.entity';
+import { IAddPromoCode, IPageOptions, IPromoCode, IPromoCodeName, IRemovePromoCode } from './promo.interface';
+
+export class PromocodeRepository {
+  async getListOfPromocodes(pageOptions: IPageOptions): Promise<PromocodeEntity[]> {
+    const skip: number = (pageOptions.page - 1) * pageOptions.limit;
+    const query = `SELECT * FROM promocode LIMIT ${pageOptions.limit} OFFSET ${skip}`;
+
+    const promocodes: PromocodeEntity[] = await appDataSource.query(query);
```
Comment (dianager) [inline src/promo/promo.repository.ts]: appDataSource should be passed using DI.

File: ormconfig.ts [hunk 17]
```diff
@@ -0,0 +1,24 @@
+import dotenv from 'dotenv';
+
```
Comment (dianager) [inline ormconfig.ts]: i think we can delete this file

File: src/promo/promo.service.ts [hunk 18]
```diff
@@ -8,52 +9,74 @@ export class PromocodeService {
...
-      throw new NotFoundException(
-          `The promocode ${promocode.name} not found.`,
-      );
+    try {
+      if (!(await this.promocodeRepository.removePromoCode(promocode))) {
+        throw new RpcException(`The promocode ${promocode.name} not found.`);
+      }
+      return true;
+    } catch (e) {
+      throw new RpcException(e);
```
Comment (dianager) [inline src/promo/promo.service.ts:38]: As far as i remember  error is type of RpcException, so we don't need to wrap it to RpcException

Comment (irnrogalik) [reply src/promo/promo.service.ts:38]: If we use RpcException, then we get an error in a readable form, otherwise just an Internal server error

**Review by dianager — APPROVED** [review]

**Pull request merged**